source/time_loop_mgrit.template.h

```diff
@@ -0,0 +1,1127 @@
+          output_quantities_multiplier_,
+          "Multiplicative modifier applied to \"output granularity\" that "
+          "determines the writeout granularity for quantities of interest");
+
+      std::copy(std::begin(HyperbolicSystemView::component_names),
+          std::end(HyperbolicSystemView::component_names),
+          std::back_inserter(error_quantities_));
+
+      add_parameter("error quantities",
+          error_quantities_,
+          "List of conserved quantities used in the computation of the "
+          "error norms.");
+
+      error_normalize_ = true;
+      add_parameter("error normalize",
+          error_normalize_,
+          "Flag to control whether the error should be normalized by "
+          "the corresponding norm of the analytic solution.");
+
+      resume_ = false;
+      add_parameter("resume", resume_, "Resume an interrupted computation");
+
+      terminal_update_interval_ = 5;
+      add_parameter("terminal update interval",
+          terminal_update_interval_,
+          "Number of seconds after which output statistics are "
+          "recomputed and printed on the terminal");
+
+      terminal_show_rank_throughput_ = true;
+      add_parameter("terminal show rank throughput",
+          terminal_show_rank_throughput_,
+          "If set to true an average per rank throughput is computed "
+          "by dividing the total consumed CPU time (per rank) by the "
+          "number of threads (per rank). If set to false then a plain "
+          "average per thread \"CPU\" throughput value is computed by "
+          "using the umodified total accumulated CPU time.");
+
+      post_process_ = false;
+      add_parameter("enable custom postproc step",
+          post_process_,
+          "do a user defined postprocessing step");
+    }
+
+//todo: past this, replace object. with object->
+    template <typename Description, int dim, typename Number>
+    void TimeLoopMgrit<Description, dim, Number>::run()
+    {
+#ifdef DEBUG_OUTPUT
+      std::cout << "TimeLoopMgrit<dim, Number>::run()" << std::endl;
+#endif
+
+      const bool write_output_files = enable_checkpointing_ ||
+          enable_output_full_ ||
+          enable_output_levelsets_;
+
+      /* Attach log file: */
+      if (mpi_rank_ == 0)
+        logfile_.open(base_name_ + ".log");
+
+      print_parameters(logfile_);
+
+      Number t = 0.;
+      unsigned int output_cycle = 0;
+      vector_type U;
+
+      /* Prepare data structures: */
+
+      const auto prepare_compute_kernels = [&]() {
+        offline_data_->prepare(problem_dimension);
+        hyperbolic_module_->prepare();
+        parabolic_module_->prepare();
+        time_integrator_->prepare();
+        postprocessor_->prepare();
+        vtu_output_->prepare();
+        /* We skip the first output cycle for quantities: */
+        quantities_->prepare(base_name_, output_cycle == 0 ? 1 : output_cycle);
+        print_mpi_partition(logfile_);
+      };
```

Comment:
see above. you could then use `level_structures.prepare()` instead of this lambda function